### PR TITLE
evidence: modify evidence types

### PR DESCRIPTION
**Pull request by cmwaters**
## Description

Changed the interface and evidence types according to [adr 59](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-059-evidence-composition-and-lifecycle.md)

### CODE REVIEW

File: types/evidence.go [hunk 1]
```diff
@@ -305,12 +136,122 @@ func DuplicateVoteEvidenceFromProto(pb *tmproto.DuplicateVoteEvidence) (*Duplica
...
+}
+
+// Hash returns the hash of the header and the commonHeight. This is designed to cause hash collisions with evidence
+// that have the same conflicting header and common height but different permutations of validator commit signatures.
+// The reason for this is that we don't want to allow several permutations of the same evidence to be committed on
+// chain. Ideally we commit the header with the most commit signatures but anything greater than 1/3 is sufficient.
+func (l *LightClientAttackEvidence) Hash() []byte {
+	bz := make([]byte, tmhash.Size+8)
+	copy(bz[:tmhash.Size-1], l.ConflictingBlock.Hash().Bytes())
+	copy(bz[tmhash.Size:], []byte(strconv.Itoa(int(l.CommonHeight))))
```
Comment (melekes) [inline types/evidence.go:185]: maybe it's better to use varint encoding here

```
buf := make([]byte, binary.MaxVarintLen64)

	for _, x := range []int64{-65, -64, -2, -1, 0, 1, 2, 63, 64} {
		n := binary.PutVarint(buf, x)
		fmt.Printf("%x\n", buf[:n])
	}
```

File: types/evidence.go [hunk 2]
```diff
@@ -305,12 +136,122 @@ func DuplicateVoteEvidenceFromProto(pb *tmproto.DuplicateVoteEvidence) (*Duplica
...
+	if l.ConflictingBlock.Header == nil {
+		return errors.New("conflicting block missing header")
+	}
+
+	if err := l.ConflictingBlock.ValidateBasic(l.ConflictingBlock.ChainID); err != nil {
+		return fmt.Errorf("invalid conflicting light block: %w", err)
+	}
+
+	if l.CommonHeight <= 0 {
+		return fmt.Errorf("incorrect common height (#%d <= 0)", l.CommonHeight)
```
Comment (melekes) [inline types/evidence.go:205]:
```suggestion
		return errors.New("negative or zero common height")
```

File: types/evidence.go [hunk 3]
```diff
@@ -338,13 +279,103 @@ func (evl EvidenceList) String() string {
...
+	tmjson.RegisterType(&DuplicateVoteEvidence{}, "tendermint/DuplicateVoteEvidence")
+	tmjson.RegisterType(&LightClientAttackEvidence{}, "tendermint/LightClientAttackEvidence")
+}
+
+//-------------------------------------------- ERRORS --------------------------------------
+
+// ErrEvidenceInvalid wraps a piece of evidence and the error denoting how or why it is invalid.
+type ErrEvidenceInvalid struct {
+	Evidence   Evidence
+	ErrorValue error
```
Comment (melekes) [inline types/evidence.go:350]:
```suggestion
	Reason error
```

File: types/evidence.go (hunk 1 above)
Comment (cmwaters) [reply types/evidence.go:185]: But then evidence hashes will be of different lengths right? Is that okay?

Comment (cmwaters) [reply types/evidence.go:185]: Ah no don't worry

**Review by melekes — APPROVED** [review]

**Pull request merged**